app/javascript/pages/UtmLinks/Index.tsx

```diff
@@ -57,107 +50,139 @@ const utmLinkWithStats = (utmLink: SavedUtmLink, stats?: UtmLinkStats) => {
   return link;
 };
 
-const UtmLinkList = () => {
-  const navigation = useNavigation();
-  const navigate = useNavigate();
-  const revalidator = useRevalidator();
-  const { utm_links: utmLinks, pagination } = cast<{ utm_links: SavedUtmLink[]; pagination: PaginationProps }>(
-    useLoaderData(),
-  );
-  const [utmLinksStats, setUtmLinksStats] = React.useState<UtmLinksStats>({});
+export const extractSortParam = (searchParams: URLSearchParams): Sort<SortKey> | null => {
+  const column = searchParams.get("key");
+  switch (column) {
+    case "link":
+    case "date":
+    case "source":
+    case "medium":
+    case "campaign":
+    case "clicks":
+    case "sales_count":
+    case "revenue_cents":
+    case "conversion_rate":
+      return {
+        key: column,
+        direction: searchParams.get("direction") === "desc" ? "desc" : "asc",
+      };
+    default:
+      return null;
+  }
+};
+
+export default function UtmLinksIndex() {
+  const {
+    utm_links: utmLinks,
+    pagination,
+    query: initialQuery,
+    sort: initialSort,
+    utm_links_stats: utmLinksStats,
+  } = usePage<UtmLinksIndexProps>().props;
+  const isNavigating = useRouteLoading();
+
   const utmLinksWithStats = utmLinks.map((utmLink) => utmLinkWithStats(utmLink, utmLinksStats[utmLink.id]));
   const [selectedUtmLink, setSelectedUtmLink] = React.useState<SavedUtmLink | null>(null);
-  const [searchParams, setSearchParams] = useSearchParams();
-  const [sort, setSort] = React.useState<Sort<SortKey> | null>(
-    () => extractSortParam(searchParams) || { key: "date", direction: "desc" },
-  );
+  const [sort, setSort] = React.useState<Sort<SortKey> | null>(() => initialSort || { key: "date", direction: "desc" });
   const [deletingUtmLink, setDeletingUtmLink] = React.useState<{
     id: string;
     title: string;
     state: "delete-confirmation" | "deleting";
   } | null>(null);
 
-  const activeStatsRequest = React.useRef<{ cancel: () => void } | null>(null);
-  const debouncedGetUtmLinksStats = useDebouncedCallback((ids: string[]) => {
-    activeStatsRequest.current?.cancel();
-    asyncVoid(async () => {
-      const request = getUtmLinksStats({ ids });
-      activeStatsRequest.current = request;
-      const stats = await request.response;
-      setUtmLinksStats((prev) => ({ ...prev, ...stats }));
-    })();
+  const fetchUtmLinksStats = useDebouncedCallback((ids: string[]) => {
+    const url = new URL(window.location.href);
+    url.searchParams.set("ids", ids.join(","));
+    router.get(url.toString(), {}, { only: ["utm_links_stats"], preserveUrl: true, preserveState: true });
```

Comment:
Would this work? `router.reload` preserves state by default
```suggestion
    router.reload({
      data: { ids: ids.join(",") },
      only: ["utm_links_stats"],
      preserveUrl: true
    });
```